theano/compile/function_module.py

```diff
@@ -121,26 +121,6 @@ def fgraph_updated_vars(fgraph, expanded_inputs):
     return updated_vars
 
 
-class Supervisor:
-    """
-    Listener for FunctionGraph events which makes sure that no
-    operation overwrites the contents of protected Variables. The
-    outputs of the FunctionGraph are protected by default.
-
-    """
-
-    def __init__(self, protected):
-        self.protected = list(protected)
-
-    def validate(self, fgraph):
```

Comment:
New way of doing it:
- start over from the faster topo branch
- Modify destroyhandler to add a new fgraph.has_destroyers(list) (as fgraph.destroyers) that will return True is any of the variable in the list have a destroyers. (the old Superviser.validate check without the raise)
- Make that new fct do the Superviser.validate check when self.algo = cycle.
- Make sure all tests still pass
- Modify the new method to be faster (don't build the big data structure) when the algo is fast.